File: src/bercpp/BERLong64Unsigned.cpp
#include "BERCodes.h"
#include "BERException/BERExceptions.h"
#include "BERLong64Unsigned.h"
#include "BasicBuffer.h"
#include <stdint.h>
#include <list>
#include <sstream>
using namespace std;

namespace BER										
{
	uint32_t BERLong64Unsigned::Size(uint64_t *value)
	{
		return 9;
	}

	BasicBuffer* BERLong64Unsigned::Encode(uint64_t *value)
	{
		BasicBuffer *result = new BasicBuffer(Size(value));
		result->PushBack(BER::Long64Unsigned);
		for (int8_t i = 7; i >= 0; i--)
		{
			result->PushBack(((*value) >> (i * 8)) & 0xFF);
		}
		return result;
	}

	uint64_t* BERLong64Unsigned::Decode(BasicBuffer &buffer, uint32_t &index)
	{
		if (buffer.Get(index) != BER::Long64Unsigned)
		{
			throw BERTypeIncorrectTag();
		}
		index++;
		uint64_t *temp = new uint64_t(0);
		for (uint8_t i = 0; i < 8; i++)
		{
			*temp = (((*temp) << 8) | buffer.Get(index++));
		}
		return temp;
	}
	
	BERLong64Unsigned::~BERLong64Unsigned()
	{
		DELETE(this->Value);
	}

	string BERLong64Unsigned::toString()
	{
		ostringstream oss;
		oss << "{BERLong64Unsigned: " << (uint64_t)(*Value) << "}" ;
		return oss.str();
	}

}

File: src/bercs/BERLong64Unsigned.cs
using System;
using System.Collections.Generic;

using BER.Exceptions;

namespace BER
{
    public class BERLong64Unsigned : BaseBERType<UInt64>
    {
        #region Overrides of BaseBERType<ulong>
		//! Method that encodes the value of this BERLong64Unsigned into a Byte List
		//! \return A Byte List containing the encoded raw bytes of this BERLong64Unsigned	
        public override List<byte> Encode(ulong value)
        {
            var result = new List<byte> { (Byte)BERCodes.Long64Unsigned };
            for (SByte i = 7; i >= 0; i--)
            {
                result.Add((Byte)((value >> (i * 8)) & 0xFF));
            }
            return result;
        }
		//! Method that decodes a List of Bytes into this BERLong64Unsigned object
		//! \return true The decoding process was succesfull
		//! \return true The decoding process was not succesfull
		//! \throws BERTypeDecodeIncorrectTag if an incompatible tag was found
        public override ulong Decode(List<byte> buffer)
        {
            var index = 0;
            UInt64 result = 0;
            if (buffer[index++] != (byte) BERCodes.Long64Unsigned)
            {
                throw new BERTypeDecodeIncorrectTag();
            }
            for (var i = 0; i < 8; i++)
            {
                result = (result << 8) | buffer[index++];
            }
            buffer.RemoveRange(0, 9);
            return result;
        }
		//! Method that provides a String representation of this BERLong64Unsigned
		//! \return The string representation of this BERLong64Unsigned
        public override string AsString()
        {
            return Value.ToString();
        }

        #endregion
    }
}
